iOS/Flows/Conversation/Cells/EmotionCircleCollectionViewLayout.swift

```diff
@@ -0,0 +1,187 @@
+    }
+
+    override func prepare() {
+        super.prepare()
+
+        guard let collectionView = self.collectionView else { return }
+
+        // Get all of the emotion attributes being managed by the animator.
+        let animatorAttributes
+        = self.animator.items(in: collectionView.bounds).compactMap { item in
+            return item as? EmotionCircleAttributes
+        }
+
+        // Keep track of the ids of items that currently exist in the data source.
+        var foundIds = Set<String>()
+
+        self.forEachIndexPath { indexPath in
+            // Get the id of the attributes and save it for later use.
+            let id = self.dataSource.getId(forItemAt: indexPath)
+
+            foundIds.insert(id)
+            // Check to see if the attribute already exists in the animator
+            if let matchingAttributes = animatorAttributes.first(where: { attributes in
+                return attributes.id == id
+            }) {
+                // The attributes were already added, but update the index path in case
+                // it's no longer valid due to inserts/deletes.
+                matchingAttributes.indexPath = indexPath
+            } else {
+                // The item's attributes aren't yet being managed by the animator. Add them now.
+                self.addEmotionAttributesToAnimator(for: indexPath, withId: id)
+            }
+        }
+
+        // Removing any attributes from the animator that no longer exists.
+        animatorAttributes.forEach { attributes in
+            guard !foundIds.contains(attributes.id) else { return }
+            self.removeDynamicItemFromAnimator(attributes)
+        }
+    }
+
+    override func layoutAttributesForElements(in rect: CGRect) -> [UICollectionViewLayoutAttributes]? {
+        return self.animator.items(in: rect).compactMap { item in
+            return item as? UICollectionViewLayoutAttributes
+        }
+    }
+
+    override func layoutAttributesForItem(at indexPath: IndexPath) -> UICollectionViewLayoutAttributes? {
+        return self.animator.layoutAttributesForCell(at: indexPath)
+    }
+
+    // MARK: - Animator Functions
+
+    
+    private func addEmotionAttributesToAnimator(for indexPath: IndexPath, withId id: String) {
+        guard let collectionView = self.collectionView,
+              collectionView.width > 0, collectionView.height > 0 else { return }
+
+        let attributes = EmotionCircleAttributes(forCellWith: indexPath)
+        attributes.id = id
+
+        let clampedDiameter = clamp(self.cellDiameter,
+                                    0,
+                                    min(collectionView.width, collectionView.height))
+        attributes.size = CGSize(width: clampedDiameter, height: clampedDiameter)
+        attributes.frame.origin
+        = CGPoint(x: CGFloat.random(in: 0...collectionView.width - clampedDiameter),
+                  y: CGFloat.random(in: 0...collectionView.height - clampedDiameter))
```

Comment:
We have to make sure that cells fit inside the physical boundaries of the view. We also have to make sure they're non-zero sized or else the animator crashes.